### PR TITLE
Adding proxy variables to FILTERED_ENV

### LINKED ISSUE #3500: https_proxy not in FILTERED_ENV
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

# Please always follow these steps:
- [x] Confirmed this is a problem with running a `brew` command and not `brew install`ing or the post-install behaviour of one or more formulae? If it's a formulae-specific problem please file this issue at the relevant tap e.g. for Homebrew/homebrew-core https://github.com/Homebrew/homebrew-core/issues/new
- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed all issues and retried your prior step?
- [x] Ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- run `brew update` behind proxy
- got `fatal: unable to access 'https://github.com/Homebrew/brew.git/': Failed to connect to github.com port 443: Operation timed out`
  - git command failed (`/Applications/Xcode.app/Contents/Developer/usr/bin/git fetch --tags --force -q origin refs/heads/master:refs/remotes/origin/master`)

# Features

- https://github.com/Homebrew/brew/blob/master/bin/brew#L91
  - there is no http**s**_proxy in FILTERED_ENV (but http_proxy)
    - see: #3494 / 63408a10ff677909f4397c5c94cfa1f2007c0826
  - please add `https_proxy`, `ftp_proxy`, `no_proxy`

We will close this issue or ask you to create a pull-request if it's something we're not actively planning to work on.

### PR DESCRIPTION
**Pull request by barreyra**
Fixes #3500

- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [X] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: bin/brew [hunk 1]
```diff
@@ -89,6 +89,7 @@ then
 
   FILTERED_ENV=()
   for VAR in HOME SHELL PATH TERM LOGNAME USER CI TRAVIS http_proxy \
+             https_proxy ftp_proxy no_proxy \
```
Comment (MikeMcQuaid) [inline bin/brew]: What is `no_proxy` used for? Can you move `http_proxy` from the previous line, thanks!

Comment (barreyra) [reply bin/brew]: `no_proxy` lets you list internal domains that should not use the proxy server. Patch updated.

Comment (MikeMcQuaid) [reply bin/brew]: Why would you use this with Homebrew?

Comment (barreyra) [reply bin/brew]: What if somebody had some private casks in their intranet, for instance?

Comment (MikeMcQuaid) [reply bin/brew]: @barreyra Which of these variables do you personally need for Homebrew to function? I don't want to whitelist everything that could be useful but just things that individuals need.

Comment (barreyra) [reply bin/brew]: I don't personally need no_proxy, but I think all others are likely needed. It just depends on what protocol is required to retrieve a given dependency (http/https/ftp).

Comment (MikeMcQuaid) [reply bin/brew]: @barreyra I'd suggest just going for `HTTPS_PROXY` (https://github.com/Homebrew/brew/issues/3500) and, if you personally use it, `FTP_PROXY` for now (both uppercase). If you also use `ALL_PROXY` or `NO_PROXY` they can be added too (but only if you personally need them).

Comment (barreyra) [reply bin/brew]: So *not* setting the http proxy?

What's the difference between the lower and uppercase versions of the variables? I've always used the lowercase versions.

Comment (barreyra) [reply bin/brew]: Answering my own question, from the curl manpage:

>The  environment  variables  can  be  specified  in lower case or upper case. The lower case version has precedence. http_proxy is an exception as it is only available in lower case.

Why do you prefer the uppercase version?

Comment (MikeMcQuaid) [reply bin/brew]: > Why do you prefer the uppercase version?

That's the version that's documented in the man page if you're copy-pasting.

Comment (barreyra) [reply bin/brew]: I'm fine going with this for now. I had added no_proxy because it was one of the asks in #3500

**Comment by MikeMcQuaid** [comment]
Thanks for your first contribution to Homebrew, @barreyra! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by barreyra** [comment]
Thanks, but it's not my first commit, I added libbluray support to ffmpeg back in the day :)

https://github.com/Homebrew/homebrew-core/commit/e451b7f62df820e39d997aaeea3da7ba59af46ba
https://github.com/Homebrew/homebrew-core/commit/d12c637c38a0bfd8c26ce2aaa38a054a3712224a

**Comment by MikeMcQuaid** [comment]
First commit to Homebrew/brew then 😁

**Pull request merged**